File: express/src/routes/Players/player_collections.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { PlayerCollectionService } from '../../services/Players/player_collection_service.js';

const router = Router();
const service = new PlayerCollectionService();

function validate(data: any): void {
  if (!((data.quantity == null || data.quantity > 0))) throw new Error(`Collection quantity must be greater than zero`);
}
function applyProjection(obj: any): any {
  if (!obj) return obj;
  const r = { ...obj };
  if ('acquiredAt' in r) { r.acquiredAt = r.acquiredAt; delete r.acquiredAt; }
  return r;
}


router.get('/', async (req, res) => {
  const items = await prisma.playerCollection.findMany();
  res.json(items.map(applyProjection));
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.quantity !== undefined) data.quantity = body.quantity;
    if (body.foil !== undefined) data.foil = body.foil;
    if (body.condition !== undefined) data.condition = body.condition;
    if (body.acquiredAt !== undefined) data.acquiredAt = body.acquiredAt != null ? new Date(body.acquiredAt) : null;
    if (body.acquiredVia !== undefined) data.acquiredVia = body.acquiredVia;
    if (body.playerId !== undefined) data.playerId = body.playerId;
    if (body.cardId !== undefined) data.cardId = body.cardId;
  try {
  validate(data);
    const entity = await prisma.playerCollection.create({ data });
    res.status(201).json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Validation error') });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.playerCollection.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  if (entity.playerId !== (req as any).userId) return res.status(403).json({ error: 'You do not own this resource.' });
  res.json(applyProjection(entity));
});

router.patch('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.quantity !== undefined) data.quantity = body.quantity;
    if (body.foil !== undefined) data.foil = body.foil;
    if (body.condition !== undefined) data.condition = body.condition;
    if (body.acquiredAt !== undefined) data.acquiredAt = body.acquiredAt != null ? new Date(body.acquiredAt) : null;
    if (body.acquiredVia !== undefined) data.acquiredVia = body.acquiredVia;
    if (body.playerId !== undefined) data.playerId = body.playerId;
    if (body.cardId !== undefined) data.cardId = body.cardId;
  try {
  validate(data);
    const existing = await prisma.playerCollection.findUnique({ where: { id: Number(req.params.id) } });
    if (!existing) return res.status(404).json({ error: 'Not found' });
  if (existing.playerId !== (req as any).userId) return res.status(403).json({ error: 'You do not own this resource.' });
    const entity = await prisma.playerCollection.update({ where: { id: Number(req.params.id) }, data });
    res.json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Error') });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    const existing = await prisma.playerCollection.findUnique({ where: { id: Number(req.params.id) } });
    if (!existing) return res.status(404).json({ error: 'Not found' });
  if (existing.playerId !== (req as any).userId) return res.status(403).json({ error: 'You do not own this resource.' });
    await prisma.playerCollection.delete({ where: { id: Number(req.params.id) } });
    res.status(204).send();
  } catch {
    res.status(404).json({ error: 'Not found' });
  }
});

router.post('/:id/add', async (req, res) => {
  const id = Number((req.params as any).id);
  const quantity = req.body.quantity;
  try {
    await service.add(id, quantity);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.get('/:id/value', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    const result = await service.estimated_value(id);
    res.json({ result });
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});
export default router;

File: gin/internal/handler/players/player_collections.go
package handler_players

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/players"
	"cards_project/internal/handler"
)

type PlayerCollectionHandler struct { db *gorm.DB }

func NewPlayerCollectionHandler(db *gorm.DB) *PlayerCollectionHandler {
	return &PlayerCollectionHandler{db: db}
}

func (h *PlayerCollectionHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/player_collections")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.DELETE("/:id", h.Delete)
	g.POST("/:id/api/collection/{id}/add", h.Add)
	g.POST("/:id/api/collection/{id}/remove", h.Remove)
	g.GET("/:id/api/collection/{id}/value", h.EstimatedValue)
}

func (h *PlayerCollectionHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.PlayerCollection
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.PlayerCollectionResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *PlayerCollectionHandler) Create(c *gin.Context) {
	var req model.PlayerCollectionCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	if msgs := validatePlayerCollection(&req); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	row := model.PlayerCollection{}
	row.Quantity = req.Quantity
	row.Foil = req.Foil
	row.Condition = req.Condition
	row.AcquiredAt = req.AcquiredAt
	row.AcquiredVia = req.AcquiredVia
	row.PlayerID = req.PlayerID
	row.CardID = req.CardID
	if err := h.db.Create(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *PlayerCollectionHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.PlayerCollection
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "PlayerCollection"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *PlayerCollectionHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.PlayerCollection
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "PlayerCollection"); return }
		handler.DbError(c, err); return
	}
	var req model.PlayerCollectionUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	createReq := toCreateRequestPlayerCollection(&row)
	if msgs := validatePlayerCollection(&createReq); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *PlayerCollectionHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *PlayerCollectionHandler) Delete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	if err := h.db.Delete(&model.PlayerCollection{}, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "PlayerCollection"); return }
		handler.DbError(c, err); return
	}
	c.Status(http.StatusNoContent)
}

func (h *PlayerCollectionHandler) Add(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.PlayerCollection
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "PlayerCollection"); return }
		handler.DbError(c, err); return
	}
	var body map[string]interface{}
	_ = c.ShouldBindJSON(&body)
	quantity := func() int {
		v, ok := body["quantity"]; if !ok { return 0 }
		f, ok := v.(float64); if !ok { return 0 }
		return int(f)
	}()
	err := row.Add(quantity)
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *PlayerCollectionHandler) Remove(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.PlayerCollection
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "PlayerCollection"); return }
		handler.DbError(c, err); return
	}
	var body map[string]interface{}
	_ = c.ShouldBindJSON(&body)
	quantity := func() int {
		v, ok := body["quantity"]; if !ok { return 0 }
		f, ok := v.(float64); if !ok { return 0 }
		return int(f)
	}()
	err := row.Remove(quantity)
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *PlayerCollectionHandler) EstimatedValue(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.PlayerCollection
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "PlayerCollection"); return }
		handler.DbError(c, err); return
	}
	result, err := row.EstimatedValue()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

func validatePlayerCollection(req *model.PlayerCollectionCreateRequest) []string {
	var errs []string
	if !(req.Quantity > 0) {
		errs = append(errs, "Collection quantity must be greater than zero")
	}
	return errs
}

func toCreateRequestPlayerCollection(m *model.PlayerCollection) model.PlayerCollectionCreateRequest {
	return model.PlayerCollectionCreateRequest{
		Quantity: m.Quantity,
		Foil: m.Foil,
		Condition: m.Condition,
		AcquiredAt: m.AcquiredAt,
		AcquiredVia: m.AcquiredVia,
		PlayerID: m.PlayerID,
		CardID: m.CardID,
	}
}
